File: src/types/booking.ts
import type { Address } from './user';
import type { Service } from './service';

export type BookingStatus =
  | 'pending'
  | 'accepted'
  | 'worker-assigned'
  | 'worker-on-the-way'
  | 'arrived'
  | 'started-work'
  | 'completed'
  | 'cancelled'
  | 'refunded';

export type PaymentStatus = 'unpaid' | 'paid' | 'refunded' | 'failed';
export type PaymentMethod = 'wallet' | 'card' | 'cash' | 'upi';

export interface BookingProblemImage {
  url: string;
  caption?: string;
}

export interface Booking {
  _id: string;
  customerId: string;
  customer?: { _id: string; name: string; avatar?: string; phone?: string };
  workerId: string;
  worker?: { _id: string; name: string; avatar?: string; profession: string };
  serviceId: string;
  service?: Service;
  status: BookingStatus;
  paymentStatus: PaymentStatus;
  paymentMethod?: PaymentMethod;
  problemDescription: string;
  problemImages: BookingProblemImage[];
  scheduledDate: string;
  scheduledTime: string;
  address: Address;
  price: number;
  currency: string;
  couponCode?: string;
  discount?: number;
  finalPrice: number;
  transactionId?: string;
  paidAt?: string;
  etaMinutes?: number;
  distanceKm?: number;
  feedbackId?: string;
  createdAt: string;
  updatedAt: string;
}

export interface CreateBookingPayload {
  serviceId: string;
  workerId: string;
  problemDescription: string;
  problemImages: string[];
  scheduledDate: string;
  scheduledTime: string;
  address: Address;
  paymentMethod: PaymentMethod;
  couponCode?: string;
}

export type BookingQuery = {
  status?: BookingStatus;
  role?: 'customer' | 'worker';
  page?: number;
  limit?: number;
};

File: backend/app/models/booking.py
"""Booking Beanie document."""
from __future__ import annotations

import enum
import uuid
from typing import Any, Dict, List, Optional

from pydantic import Field

from app.models.base import BaseDocument


class BookingStatus(str, enum.Enum):
    PENDING = "pending"
    ACCEPTED = "accepted"
    WORKER_ASSIGNED = "worker-assigned"
    WORKER_ON_THE_WAY = "worker-on-the-way"
    ARRIVED = "arrived"
    STARTED_WORK = "started-work"
    COMPLETED = "completed"
    CANCELLED = "cancelled"
    REFUNDED = "refunded"


class PaymentStatus(str, enum.Enum):
    UNPAID = "unpaid"
    PAID = "paid"
    REFUNDED = "refunded"
    FAILED = "failed"


class PaymentMethod(str, enum.Enum):
    WALLET = "wallet"
    CARD = "card"
    CASH = "cash"
    UPI = "upi"


class BookingType(str, enum.Enum):
    SCHEDULED = "scheduled"
    INSTANT = "instant"
    EMERGENCY = "emergency"


class Booking(BaseDocument):
    id: str = Field(default_factory=lambda: str(uuid.uuid4()))
    customer_id: str
    worker_id: Optional[str] = None
    service_id: Optional[str] = None
    status: BookingStatus = BookingStatus.PENDING
    payment_status: PaymentStatus = PaymentStatus.UNPAID
    payment_method: Optional[PaymentMethod] = None
    problem_description: str
    problem_images: List[str] = Field(default_factory=list)
    scheduled_date: str
    scheduled_time: str
    address: Dict[str, Any]
    price: float
    currency: str = "INR"
    coupon_code: Optional[str] = None
    discount: float = 0.0
    final_price: float
    transaction_id: Optional[str] = None
    paid_at: Optional[str] = None
    refunded_at: Optional[str] = None
    refund_reason: Optional[str] = None
    eta_minutes: Optional[int] = None
    distance_km: Optional[float] = None
    booking_type: BookingType = BookingType.SCHEDULED
    is_emergency: bool = False

    class Settings:
        name = "bookings"
